### DISCUSSION TITLE
Help installing cloned pymeasure repo with pip

### DISCUSSION THREAD

**Discussion by jwbrooks0** [pymeasure/pymeasure#893] [Q&A] [answered]
I'm working through the instructions [here](https://pymeasure.readthedocs.io/en/latest/dev/contribute.html) and I get to the instructions that say:
`cd /path/for/code/pymeasure`
`pip install -e .`

When I run `pip install`, it complains that there is no `setup.py` file.  I checked, and there is in fact no `setup.py` file anywhere in the repo.  Am I missing something??  

As an alternate method, I tried to add `path/for/code/pymeasure` to my PYTHONPATH (the usual way I "install" repositories), but I get the following error when I try to import.  
`XXXXX/GitHub/pymeasure/pymeasure/__init__.py:49: UserWarning: Could not find pymeasure version, it does not seem to be installed. Either install it (editable or full) or install setuptools_scm
  warnings.warn('Could not find pymeasure version, it does not seem to be installed. `

I'm by no means a python expert so maybe I'm missing something.  

I'm using GitHub Desktop, virtualenv, python 3.8.10.,  Ubuntu 20.04.6 LTS

**Comment by jwbrooks0** [comment]
After more digging, [this website](https://towardsdatascience.com/setuptools-python-571e7d5500f2#:~:text=you%20can%20now%20create%20a%20dummy) recommended creating a _dummy_ setup.py file.  I did so, and the install seems to have worked.  

Is this the way it is supposed to work?  Should the contributing website be updated?  And why is it setup so I can't add the pymeasure directory to my PYTHONPATH?

**Reply by bilderbuchi** [reply]
> Is this the way it is supposed to work?

No

> Should the contributing website be updated?

No, setup.py is the legacy way, and will stop working at some point

> And why is it setup so I can't add the pymeasure directory to my PYTHONPATH?

You can add it to `PYTHONPATH` (and you did).
Just at import time the fact that this is not a full "installation" surfaces. Installing setuptools_scm as advised might help. (This is needed so we can report an as-precise-as-possible pymeasure version, as requested by users.

**Comment by bilderbuchi** [comment]
> When I run pip install, it complains that there is no setup.py file. I checked, and there is in fact no setup.py file anywhere in the repo. Am I missing something??

Yes, your pip and/or setuptools is probably too old - installation without the legacy setup.py has been added quite some time ago (I don't know the exact version). I cannot reproduce this with pip 23.1, ` pip install -e .` completes as expected.

>  Either install it (editable or full) or install setuptools_scm 

Have you tried installing `setuptools_scm` as the message advises?

**Comment by jwbrooks0** [comment]
@bilderbuchi , Thanks very much for the detailed reply!  This is the first I've heard of this new way to install repos.  

> Have you tried installing setuptools_scm as the message advises?

I tried related google searches (how I came across the dummy setup.py solution) but didn't follow that particular thread.

**Reply by bernhardlang** [reply, marked as answer]
I ran into the same issue.

~$ pip install -e .
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode
(A "pyproject.toml" file was found, but editable mode currently requires a setup.py based build.)

pip was version 20.3.4. 

~$ python -m pip install --upgrade pip

upgraded pip to 23.3 and solved the issue.

**Comment by seb5g** [comment]
Hello,
I actually run into another issue when trying to install from the cloned repository.

If you do so: 
```
pip install -e .
```

and if your shell is not aware of git (which was my case) then the installation complains it cannot find the version because it cannot find git. It troubled me for quite some time before I realized this. So if you install using pip the repo code (from a zip) of from a clone but with unaware git app, the installation will fail...

**Reply by BenediktBurger** [reply]
In your cases (zip from Github / no git / using pymeasure as a git submodule...) you can set the environment variable `SETUPTOOLS_SCM_PRETEND_VERSION` to some version you desire.
Afterwards the install will succeed (with that version number assigned) .
